capnp/src/introspect.rs

```diff
@@ -213,8 +213,9 @@ pub struct RawStructSchema {
 
     /// Map from discriminant value to field index.
     pub members_by_discriminant: &'static [u16],
-    //
-    // TODO: members_by_name, allowing fast field lookup by name.
+
+    // Sorted array of field names with matching field index.
+    pub members_by_name: &'static [(&'static str, u16)],
```

Comment:
The direct translation of capnproto-c++ would make this a `&[u16]`: https://github.com/capnproto/capnproto/blob/537fe97d6ab1962e92e15b8a16f60439a63b90f5/c%2B%2B/src/capnp/raw-schema.h#L184

Is there a reason that you did it this way? The `&str` values that you are storing here should be cheaply retrievable via `StructSchema::get_fields()`.

Reply:
Did not know it was cheap to get that, thanks for the tip. I saw some calls to `Reader::get_text` in the benchmarking flamegraph that seemed expensive, but I think that was only because of the total number of calls to that function (which we already reduced).

Only problem now is that we don't have `PartialOrd` to compare the field name passed by the user and `text::Reader`.

Reply:
I tested on [this branch](https://github.com/capnproto/capnproto-rust/compare/master...vectra-ai-research:capnproto-rust:get-name) and using `StructSchema::get_fields` made it slower than the O(N) implementation. Most of the time is spent getting the `FieldList` followed by `get_name`.

Reply:
In that branch, you are converting to `utf8` before making a comparison:
```rust
} else if candidate_name.to_str()? < name {
```
If you directly make the comparison on `capnp::text::Reader`, how much faster does that make things? (Implementing `PartialOrd` for `text::Reader` should be easy, because `&[u8]` already implements `PartialOrd`.)

It might also help to add `#[inline]` attributes on various methods, like `StructSchema::get_proto()`. (I don't particularly expect this to help, unless it somehow enables cross-crate inlining, but it seems worth a try.)

Reply:
I added the `PartialOrd` trait to the for the `capnp::text::Reader` but the less than comparison with `&str` does not compile. The compiler complains that they are different types. I was surprised since the equal to comparison compiles. For what it is worth the `to_str` call adds a negligible amount to the flamegraph compared to `get_name`. I obviously am seeking performance improvement but ideally not adding any unnecessary complexity that makes it more confusing to read.

I will try inlining `StructSchema::get_proto`. I was also considering an `Option` field in the `StructSchema` to cache the `FieldList` after the first call to `StructSchema::get_fields`.

Reply:
> but the less than comparison with &str does not compile... I was surprised since the equal to comparison compiles

Here are the impls for `PartialEq` which make that possible:

https://github.com/capnproto/capnproto-rust/blob/e476eaaeeee8ef3c079782d7893547c8e9a2aee0/capnp/src/text.rs#L48-L58

I think something similar can be done with `PartialOrd`.

Reply:
added in f6adc78bcf892f4bab9bea5cecf2e6bb6a3627ac

Reply:
Neither change improved performance. I have attached a screenshot of the latest flamegraph. The `ParitalOrd` takes as long as the `to_str` comparison.

![Screenshot 2024-01-08 at 9 55 50 AM](https://github.com/capnproto/capnproto-rust/assets/8449196/03f944c6-fd80-40f9-a38c-42b399884793)

Reply:
Thanks for the flame graph.

To summarize:
 1. Status quo is that `get_field_by_name` accounts for 43% of your running time.
 2. Implementing binary search on `members_by_name: &'static [u16]` brings that down to 29%.
 3. Implementing binary search on `members_by_name: &'static [(&'static str, u16)]` brings it down to 19%.
 
The cost of (3) is that it increases the size of the generated code, because then fields names get included twice (once in the serialized schema, and once in this new array). Moreover, I'm resistant to diverging from the capnproto-c++ way of doing things.

---

Other `#[inlines]` could potentially help, like maybe `FieldList::get()` and `struct_list::Reader::get()`.

Is there an easy way that you could share a benchmark with me so that I can play around with it?

Reply:
@quartox roughly how many fields do your structs have?

Reply:
I was going to try to share the benchmark but it isn't as simple as I had hoped. I was going to construct a schema with similar properties and generate fake data. I haven't had much success using the command line capnp json to binary converter with as much nesting as is happening in my real schemas. I will see if I can find another way to do this.

I will clean up this PR to use `members_by_name: &'static [u16]` and then we can open another issue to discuss the other performance improvements. Ultimately speeding up the other portions of the code would be more beneficial.

I may explore a static hashmap since the C++ implementation has a todo to make that improvement, but if you want to match what they currently do then we may not adopt it.

Reply:
I'm considering making a version-bump release soon, including #472 as well.

I like the `members_by_name: &'static [u16]` solution. Can you submit a pull request for that?

Reply:
Ah, I should have noticed that! That should be sufficient. Thanks!